### PR TITLE
Optimize mocking further by retaining mocks across sessions

### PR DESCRIPTION
**Pull request by tomas-abrahamsson**
This takes the optimizations a bit further by retaining mocked modules between tests in a `?WITH_MOCKED_SETUP`. That is, between such tests, the mocked modules are not restored to original, but instead the modules are left in a mocked state, the next test will need them mocked.

For a fairly large unit test module with 44 test cases with many mocked functions (the same as in #18) the timing now looks like in the table below (lower value is better).

Execution time  | Code
----------------|----------------
67s             | Original
32s             | PR #18
9s              | This PR

This PR currently shares its first commits with that of #18, I didn't know how to do it before that one gets merged, yet I thought it may be worth to make a PR so any discussions can happen. I can rebase this later on when it is suitable. I suppose it may make sense to have this further optimization as a separate PR instead of baking it into the #18, but I'm open to any approach, just let me know.

The general optimization idea in this PR is as follows:

* Extend [`mockgyver:exec/3→4`](https://github.com/tomas-abrahamsson/mockgyver/blob/retain-mocks-across-sessions/src/mockgyver.erl#L438-L462) with an extra parameter for options
* In `?WITH_MOCKED_SETUP`, set a `mock_sequence` option to tell mockgyver that the session is part of a longer mock sequence.
* When a session in a sequence ends, don't restore mocked modules, instead await the next session.
* But when the _last_ session in the sequence ends, _do_ restore mocked modules.
* When a session in a sequence starts, don't re-setup mocked modules.
* Use a test signature to tell different mock sequences apart. This signature consists of the current location and some more to make it unique for each `?WITH_MOCKED_SETUP`.
* When awaiting the next sequence, if it doesn't started quickly enough, currently 100ms, then restore the mocked modules. Something might be wrong. If we this turned out to be overly cautious, the only thing that happened was that the optimization didn't kick in. It might kick in for the next session.
* With just `?MOCK` as opposed to `?WITH_MOCKED_SETUP`, don't set any mock sequence options.
* There is a `no_mock_sequence` option to disable this optimization.

So modules stay mocked even after a `mockgyver:exec` session. But only between test in `?WITH_MOCKED_SETUP`. And with a safety timer.

This PR introduces a backwards incompatible change: it no longer works to mock the `application` module with `?WITH_MOCKED_SETUP`, since the `mockgyver:exec/4` depends on it (uses it). The retaining of mockings between sessions collides with possibilities of mocking the `application` module. I figure it is a not so good practice to mock modules central to Erlang/OTP, so I hope this is not too big of a draw-back. (I don't think it will work even if the mocking is done carefully to fall back to the renamed original: the mocking module requires a session but when calling `mockgyver:exec/4`, we are between sessions. The `no_mock_sequence` option may be used though.)

Of course, code review and comments are welcome.

**Edit:** Forgot to say that I have tested this in a large project. Only small adjustments were needed. For instance one test didn't properly wait for a process to terminate, and since the tests now run faster, there was a `{error,{already_started,<pid>}}` false positive.

### CODE REVIEW

File: include/mockgyver.hrl [hunk 1]
```diff
@@ -5,22 +5,53 @@
...
+                StackId = erlang:phash2(?mockgyver_get_stacktrace__()),
+                Base = #{num_sessions => length(Tests),
+                         signature => {?FILE, ?LINE, ?FUNCTION_NAME, StackId}},
+                IndexedTests = lists:zip(lists:seq(1, length(Tests)),
+                                         Tests),
+                {timeout, ForAllTimeout,           %% timeout for all tests
+                 [{timeout, PerTcTimeout,          %% timeout for each test
+                   {spawn,
+                    {atom_to_list(__Test),         %% label per test
+                     fun() ->
```
Comment (tomas-abrahamsson) [inline include/mockgyver.hrl:36]: This no longer uses `?WITH_FUN` from eunit_addons due to the need for indexing. Is it desirable to pursue somehow extending eunit_addons to cover this usecase? if yes, then how?

File: include/mockgyver.hrl [hunk 2]
```diff
@@ -5,22 +5,53 @@
...
+-define(mockgyver_get_stacktrace__(),
+        try error(get_stack)
+        catch error:get_stack:St -> St
+        end).
+-else. % OTP_RELEASE
+-define(mockgyver_get_stacktrace__(),
+        try error(get_stack)
+        catch error:get_stack -> erlang:get_stacktrace()
+        end).
+-endif. % OTP_RELEASE.
```
Comment (tomas-abrahamsson) [inline include/mockgyver.hrl:18]: Similar question about what Erlang version to aim for. For now, it is the same as in #18, but it is easy to change.

Comment (klajo) [reply include/mockgyver.hrl:18]: As in #18 is good. I just merged that one as is.

File: include/mockgyver.hrl (hunk 1 above)
Comment (klajo) [reply include/mockgyver.hrl:36]: There's no self-interest in using eunit_addons as such, but I'll have a deeper look at the changes.

**Comment by tomas-abrahamsson** [comment]
I've rebased on top of latest master and force-pushed.

**Comment by klajo** [comment]
> I've rebased on top of latest master and force-pushed.

Thanks!

I'll have a closer look and get back to you.

**Comment by tomas-abrahamsson** [comment]
I added a commit, to provide a better error message to the user/programmer in case some central module is mocked, in particular something that is called by eunit or mockgyver itself between sessions in a sequence. The error message recomments trying to avoid mocking that module, or else to use the `no_mock_sequence` option to the `?WITH_MOCKED_SETUP`.

File: test/mockgyver_tests.erl [hunk 3]
```diff
@@ -536,6 +522,199 @@ can_mock_a_module_with_on_load_aux() ->
...
+          end,
+          [{mock_sequence, #{num_sessions => 2, index => 1, signature => a}}]),
+    %% It should stay mocked across sessions in a mock_sequence,
+    %% that's the optimizaion. Can't call it to  test it though---that
+    %% would require a session---so test it in another way.
+    ?assertNotEqual(UnmockedChecksum5, mockgyver_dummy5:module_info(md5)),
+    ?MOCK(fun() ->
+                  ?WHEN(mockgyver_dummy5:a(_) -> mocked_1_2),
+                  mocked_1_2 = mockgyver_dummy5:a(1)
+          end,
```
Comment (tomas-abrahamsson) [inline test/mockgyver_tests.erl:559]:
```suggestion
    ?MOCK(fun() ->
                  1 = mockgyver_dummy5:a(1),
                  ?WHEN(mockgyver_dummy5:a(_) -> mocked_1_2),
                  mocked_1_2 = mockgyver_dummy5:a(1)
          end,
```

**Comment by tomas-abrahamsson** [comment]
I force-pushed an extra line of unit test. Now it is ready for review. Sorry for the noise. (I intended to start another review for this, just to make it clear what I was about to do, but messed up, and in the end, it just got force-pushed.)

**Comment by tomas-abrahamsson** [comment]
Hmm.. revisiting again. In another corpus of unit tests, I've spotted mockings of `filelib` and `timer`, both of which could potentially be called by eunit between unit tests in a generator, meaning letting mocking modules stay loaded could prove problematic. 

Maybe need to implement some auto-disabling of the optimization depending on what modules are mocked to make sure existing tests still work. (Still warranted, I should think, given the speedup(?))

**Comment by tomas-abrahamsson** [comment]
I just force-pushed a change to the last commit, and I think it is (really) ready for review now. (I will also try to run some more extensive tests in a big project, meanwhile.)

The situation with mocked functions possibly getting called between unit tests (by eunit, mockgyver or the rest of the runtime) was bugging me, but I think I found a nice and clean solution for it now: In the `await_next_session` state, just return `{ok,undefined}` to any `reg_call_and_get_action`, this will in effect redirect the call to the renamed `module^` and the original code will get invoked.

(I've kept a dead-end on the `https://github.com/tomas-abrahamsson/mockgyver/tree/cul-de-sac/disable-mock-seq-optimization` in the dubious case it will still be needed. On this branch, it basically disables the mock-sequence optimization if any module of any of the `kernel`, `stdlib` or `sasl` apps is mocked, and also prints a warning about it. I will delete this branch in a while if will not be needed.)

**Comment by klajo** [comment]
Marking this as closed, in favor of #21. Thanks for your hard work.

**Pull request closed without merging**